### PR TITLE
Make authorship soft deadline configurable.

### PR DESCRIPTION
**Pull request by tomusdrw**
A follow up on discussion in https://github.com/paritytech/substrate/pull/9663#issuecomment-950394269

The idea is to expose some public functions so that the soft deadline can actually be configured as a percentage of the hard deadline. Before it was hardcoded to 50%, now it's just the default and can be changed by the downstream projects.

This is especially useful for parachains, which have a pretty low absolute deadline (333ms). Note however that changing this parameter has security implications and setting it too high might end up opening a DoS vector, where block producers are forced to create & push empty blocks.

### CODE REVIEW

**Review by shawntabrizi — APPROVED** [review]

**Comment by crystalin** [comment]
@tomusdrw thank you for taking care of this !! :)

File: client/basic-authorship/src/basic_authorship.rs [hunk 1]
```diff
@@ -95,12 +106,29 @@ impl<A, B, C> ProposerFactory<A, B, C, DisableProofRecording> {
...
 			default_block_size_limit: DEFAULT_BLOCK_SIZE_LIMIT,
+			soft_deadline_percent: DEFAULT_SOFT_DEADLINE_PERCENT,
 			telemetry,
 			client,
 			include_proof_in_block_size_estimation: false,
 			_phantom: PhantomData,
 		}
 	}
+
+	/// Set soft deadline percentage.
```
Comment (bkchr) [inline client/basic-authorship/src/basic_authorship.rs:117]: This needs to be moved into the impl block below in line 177.

**Review by bkchr — CHANGES_REQUESTED** [review]

File: client/basic-authorship/src/basic_authorship.rs (hunk 1 above)
Comment (tomusdrw) [reply client/basic-authorship/src/basic_authorship.rs]: Thanks! Done.

**Review by bkchr — APPROVED** [review]

**Review by librelois — APPROVED** [review]

**Comment by librelois** [comment]
@tomusdrw we are really interested in this PR, do you know if it will be merged soon?

**Comment by tomusdrw** [comment]
bot merge

**Comment by bkchr** [comment]
bot merge

**Pull request merged**